Source/WebInspectorUI/UserInterface/Workers/HeapSnapshot/HeapSnapshot.js

```diff
@@ -761,11 +773,26 @@ HeapSnapshot = class HeapSnapshot
             pathsBeingProcessed[i] = undefined;
 
             if (this._nodeOrdinalIsGCRoot[nodeOrdinal]) {
-                let fullPath = currentPath.slice();
                 let nodeIndex = nodeOrdinal * this._nodeFieldCount;
-                fullPath.push({node: nodeIndex});
-                gcRootPaths.push(fullPath);
-                continue;
+
+                let rootReason = ""
+
+                for (let rootIndex = 0; rootIndex < this._roots.length; rootIndex += rootFieldCount) {
+                    let nodeID = snapshot._roots[rootIndex + rootNodeIdOffset]
+                    let reason = snapshot._labels[snapshot._roots[rootIndex + rootReasonIndexOffset]]
+                    if (this._nodeIdentifierToOrdinal.get(nodeID) != nodeOrdinal)
+                        continue
+                    rootReason = reason
+                    break
+                }
+
+                currentPath.push({node: nodeIndex, rootReason});
+
+                // These aren't really roots. They will only mark things that are already reachable.
+                if (rootReason != "WeakHandles") {
+                    gcRootPaths.push(currentPath);
```

Comment:
we might still need the `currentPath.slice()` as otherwise we could be saving the value which may get modified in a subsequent iteration